### PR TITLE
Simplify handling of sockaddr information lookup

### LINKED ISSUE #8: Create a class to handle `ucs_sock_addr_t`
Currently the implementation to handle [`ucs_sock_addr_t`](https://github.com/rapidsai/ucxx/blob/e1c999df4db072d036ecf157a1134335ccae0019/cpp/src/utils/sockaddr.cpp) is almost a verbatim copy of the Cython implementation from UCX-Py. This could be improved by making that a class to manage the entire lifetime of the object instead of relying on raw pointers and external destructors.

### PR DESCRIPTION
**Pull request by wence-**
Migrate away from using `gethostbyname` (deprecated) to `getaddrinfo`. As a consequence, we can refactor some memory management and handle allocation lifetime with a `unique_ptr`.

Although this doesn't introduce a class for managing sockaddr objects, it does make the lifetime management transparent enough that this closes #8.

### CODE REVIEW

File: cpp/include/ucxx/utils/sockaddr.h [hunk 1]
```diff
@@ -4,36 +4,26 @@
...
+ * @brief Get an addrinfo struct corresponding to an address and port.
  *
- * Set a socket address and port as defined by the user in a socket address storage that
- * may later be used to specify an address to bind a UCP listener to.
+ * This information can later be used to bind a UCP listener or endpoint.
  *
- * @param[in] sockaddr    pointer to the UCS socket address storage.
  * @param[in] ip_address  valid socket address (e.g., IP address or hostname) or NULL as a
  *                        wildcard for "all" to set the socket address storage to.
  * @param[in] port        port to set the socket address storaget to.
```
Comment (pentschev) [inline cpp/include/ucxx/utils/sockaddr.h:21]:
```suggestion
 * @param[in] port        port to set the socket address storage to.
```

File: cpp/src/endpoint.cpp [hunk 2]
```diff
@@ -61,17 +55,16 @@ std::shared_ptr<Endpoint> createEndpointFromHostname(std::shared_ptr<Worker> wor
...
     throw ucxx::Error("Worker not initialized");
 
-  auto params = std::unique_ptr<ucp_ep_params_t, EpParamsDeleter>(new ucp_ep_params_t);
-
-  struct hostent* hostname = gethostbyname(ipAddress.c_str());
-  if (hostname == nullptr) throw ucxx::Error(std::string("Invalid IP address or hostname"));
+  ucp_ep_params_t params{};
+  params.field_mask = UCP_EP_PARAM_FIELD_FLAGS | UCP_EP_PARAM_FIELD_SOCK_ADDR |
+                      UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE | UCP_EP_PARAM_FIELD_ERR_HANDLER;
+  params.flags = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
```
Comment (pentschev) [inline cpp/src/endpoint.cpp:61]:
```suggestion
  ucp_ep_params_t params = {.field_mask = UCP_EP_PARAM_FIELD_FLAGS | UCP_EP_PARAM_FIELD_SOCK_ADDR |
                                          UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE |
                                          UCP_EP_PARAM_FIELD_ERR_HANDLER,
                            .flags = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER};
```

File: cpp/src/endpoint.cpp [hunk 3]
```diff
@@ -81,14 +74,13 @@ std::shared_ptr<Endpoint> createEndpointFromConnRequest(std::shared_ptr<Listener
...
-  auto params        = std::unique_ptr<ucp_ep_params_t, EpParamsDeleter>(new ucp_ep_params_t);
-  params->field_mask = UCP_EP_PARAM_FIELD_FLAGS | UCP_EP_PARAM_FIELD_CONN_REQUEST |
-                       UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE | UCP_EP_PARAM_FIELD_ERR_HANDLER;
-  params->flags        = UCP_EP_PARAMS_FLAGS_NO_LOOPBACK;
-  params->conn_request = connRequest;
+  ucp_ep_params_t params{};
+  params.field_mask = UCP_EP_PARAM_FIELD_FLAGS | UCP_EP_PARAM_FIELD_CONN_REQUEST |
+                      UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE | UCP_EP_PARAM_FIELD_ERR_HANDLER;
+  params.flags        = UCP_EP_PARAMS_FLAGS_NO_LOOPBACK;
+  params.conn_request = connRequest;
```
Comment (pentschev) [inline cpp/src/endpoint.cpp:81]:
```suggestion
  ucp_ep_params_t params = {
    .field_mask = UCP_EP_PARAM_FIELD_FLAGS | UCP_EP_PARAM_FIELD_CONN_REQUEST |
                  UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE | UCP_EP_PARAM_FIELD_ERR_HANDLER,
    .flags        = UCP_EP_PARAMS_FLAGS_NO_LOOPBACK,
    .conn_request = connRequest};
```

File: cpp/src/endpoint.cpp [hunk 4]
```diff
@@ -100,12 +92,12 @@ std::shared_ptr<Endpoint> createEndpointFromWorkerAddress(std::shared_ptr<Worker
...
     throw ucxx::Error("Address not initialized");
 
-  auto params        = std::unique_ptr<ucp_ep_params_t, EpParamsDeleter>(new ucp_ep_params_t);
-  params->field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS | UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE |
-                       UCP_EP_PARAM_FIELD_ERR_HANDLER;
-  params->address = address->getHandle();
+  ucp_ep_params_t params{};
+  params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS | UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE |
+                      UCP_EP_PARAM_FIELD_ERR_HANDLER;
+  params.address = address->getHandle();
```
Comment (pentschev) [inline cpp/src/endpoint.cpp:98]:
```suggestion
  params->address        = address->getHandle();
  ucp_ep_params_t params = {.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS |
                                          UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE |
                                          UCP_EP_PARAM_FIELD_ERR_HANDLER,
                            .address = address->getHandle()};
```

**Review by pentschev — COMMENTED** [review]
Thanks @wence- , this looks great! I left a few suggestions on style, and one typo fix (which is not even modified in this PR but I catched and suggested it anyway).

**Comment by wence-** [comment]
Thanks, done!

**Comment by wence-** [comment]
I just squashed the remaining few places where param structs were separated from initialisation.

**Pull request merged**